src/ChangeStream/ChangeStream.php

```diff
@@ -0,0 +1,156 @@
+     * Create a change stream.
+     */
+    public function __construct()
+    {
+        $this->stack = array();
+
+        $this->normalizers = array(
+            new LinkResourceNormalizer(),
+            new DirectoryResourceNormalizer(),
+            new FileResourceNormalizer(),
+            new GenericResourceNormalizer(),
+        );
+    }
+
+    /**
+     * Store a version of a resource in the ChangeStream to retrieve it if needed.
+     *
+     * @param string       $path
+     * @param PuliResource $resource
+     */
+    public function log($path, PuliResource $resource)
+    {
+        if (!array_key_exists($path, $this->stack)) {
+            $this->stack[$path] = array();
+        }
+
+        $this->stack[$path][] = $this->normalize($resource);
+    }
+
+    /**
+     * Create a stack of resources for the given path.
+     *
+     * @param ResourceRepository $repository
+     * @param string             $path
+     *
+     * @return ResourceStack
+     */
+    public function buildResourceStack(ResourceRepository $repository, $path)
+    {
+        $stack = array();
+
+        if (isset($this->stack[$path]) && is_array($this->stack[$path])) {
+            foreach ($this->stack[$path] as $data) {
+                $resource = $this->denormalize($data);
+                $resource->attachTo($repository, $path);
+
+                $stack[] = $resource;
+            }
+        }
+
+        return new ResourceStack($stack);
+    }
+
+    /**
+     * @return array
+     */
+    public function getLogStack()
+    {
+        return $this->stack;
+    }
+
+    /**
+     * Register a given normalizer for usage by this change stream.
+     *
+     * @param ResourceNormalizer $normalizer
+     */
+    public function registerNormalizer(ResourceNormalizer $normalizer)
+    {
+        array_unshift($this->normalizers, $normalizer);
+    }
+
+    /**
+     * @param PuliResource $resource
+     *
+     * @return array
+     */
+    private function normalize(PuliResource $resource)
+    {
+        foreach ($this->normalizers as $normalizer) {
+            if ($normalizer->supports($resource)) {
+                $data = $normalizer->normalize($resource);
+                $data['normalizer'] = get_class($normalizer);
+
+                return $data;
+            }
+        }
+
+        throw new \RuntimeException(sprintf(
+            'Instances of "%s" are not supported by this ChangeStream as no normalizer in the '.
+            'ChangeStream supports them (see the documentation to create your own normalizer).',
+            get_class($resource)
+        ));
+    }
+
+    /**
+     * @param array $data
+     *
+     * @return PuliResource
+     */
+    private function denormalize($data)
+    {
+        foreach ($this->normalizers as $normalizer) {
+            if (get_class($normalizer) === $data['normalizer']) {
+                return $normalizer->denormalize($data);
+            }
+        }
+
+        throw new \RuntimeException(sprintf(
```

Comment:
That's something we definitely should do, but then we should do it everywhere. For the moment, there are quite a lot of places where we didn't add the exceptions in the phpdoc.